include/vara/Feature/ConstraintParser.h

```diff
@@ -0,0 +1,353 @@
+    case 'W':
+    case 'X':
+    case 'Y':
+    case 'Z':
+      return munchIdentifier(Pos);
+    default:
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::ERROR,
+                  std::string(1, Cnt[Pos]))),
+              1};
+    }
+  }
+
+  ResultTy munchWhitespace(const int Pos) {
+    int MunchLength = 0;
+    for (int LL = Pos; LL < Cnt.size(); LL++) {
+      auto Head = Cnt[LL];
+      if (Head == ' ' || Head == '\t' || Head == '\n' || Head == '\r') {
+        MunchLength++;
+      } else {
+        break;
+      }
+    }
+    return {std::move(std::make_unique<ConstraintToken>(
+                ConstraintToken::ConstraintTokenKind::WHITESPACE)),
+            MunchLength};
+  }
+
+  ResultTy munchOperator(const int Pos) {
+    switch (Cnt[Pos]) {
+    case '+':
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::PLUS)),
+              1};
+    case '-':
+      if (Pos + 1 < Cnt.size() && Cnt[Pos + 1] == '>') {
+        return {std::move(std::make_unique<ConstraintToken>(
+                    ConstraintToken::ConstraintTokenKind::IMPLIES)),
+                2};
+      }
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::MINUS)),
+              1};
+    case '<':
+      if (Pos + 2 < Cnt.size() && Cnt[Pos + 1] == '-' && Cnt[Pos + 2] == '>') {
+        return {std::move(std::make_unique<ConstraintToken>(
+                    ConstraintToken::ConstraintTokenKind::EQUIVALENT)),
+                3};
+      }
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::LESS)),
+              1};
+    case '>':
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::GREATER)),
+              1};
+    case '!':
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::NOT)),
+              1};
+    case '~':
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::NEG)),
+              1};
+    case '*':
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::STAR)),
+              1};
+    case '&':
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::AND)),
+              1};
+    case '|':
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::OR)),
+              1};
+    default:
+      return {std::move(std::make_unique<ConstraintToken>(
+                  ConstraintToken::ConstraintTokenKind::ERROR,
+                  std::string(1, Cnt[Pos]))),
+              1};
+    }
+  }
+
+  ResultTy munchNumber(const int Pos) {
```

Comment:
With StringRef this is a substring search + llvm::to_integer(), would make it a bit easier than handling everything our selfs.